src/main/scala/com/amazon/deequ/anomalydetection/seasonal/HoltWinters.scala

```diff
@@ -0,0 +1,233 @@
+
+    // mean of second period - mean of first period
+    val trend = ListBuffer(
+      (secondPeriodSum - firstPeriodSum) / (periodicity.toDouble * periodicity.toDouble)
+    )
+
+    // first `periodicity` data points - estimated level
+    val seasonality = ListBuffer(series.take(periodicity).map(_ - level.head): _*)
+    // running signal estimate
+    val y = ListBuffer(level.head + trend.head + seasonality.head)
+    // input `series`, `numberOfPointsToForecast` forecasts will be appended here
+    val Y = ListBuffer(series: _*)
+
+    for (t <- 0 until series.size + numberOfPointsToForecast) {
+      if (t >= series.size) {
+        Y.append(level.last + trend.last + seasonality(seasonality.size - periodicity))
+      }
+
+      level.append(alpha * (Y(t) - seasonality(t)) + (1 - alpha) * (level(t) + trend(t)))
+      trend.append(beta * (level(t + 1) - level(t)) + (1 - beta) * trend(t))
+      seasonality.append(gamma * (Y(t) - level(t) - trend(t)) + (1 - gamma) * seasonality(t))
+
+      y.append(level(t + 1) + trend(t + 1) + seasonality(t + 1))
+    }
+
+    // forecast residuals
+    val residuals = y.zip(series).map { case (modelForecast, seriesValue) =>
+      seriesValue - modelForecast
+    }
+    val forecasted = Y.drop(series.size)
+
+    ModelResults(forecasted, level, trend, seasonality, residuals)
+  }
+
+  private def modelSelectionFor(
+      dataSeries: Seq[Double],
+      numberOfObservationsToForecast: Int)
+    : (Double, Double, Double) = {
+
+    // solver with parameter bounds
+    val solver = new LBFGSB(
+      lowerBounds = DenseVector[Double](0, 0, 0),
+      upperBounds = DenseVector[Double](1, 1, 1)
+    )
+
+    // initial smoothing parameters for level, trend, and seasonality respectively
+    val initialParameters = DenseVector[Double](0.3, 0.1, 0.1)
+
+    val objective = new DiffFunction[DenseVector[Double]] {
+      override def calculate(x: DenseVector[Double]): (Double, DenseVector[Double]) = {
+        val modelResults = additiveHoltWinters(
+          dataSeries, seriesPeriodicity, numberOfObservationsToForecast,
+          alpha = x(0), beta = x(1), gamma = x(2)
+        )
+        val rss = modelResults.residuals.map(math.pow(_, 2)).sum
+        val gradient = DenseVector.zeros[Double](x.length)
+
+        rss -> gradient
+      }
+    }
+
+    val objectiveWithApproximateGradients =
+      new ApproximateGradientFunction[Int, DenseVector[Double]](objective)
+
+    val optimalParameters = solver.minimize(objectiveWithApproximateGradients, initialParameters)
+    (optimalParameters(0), optimalParameters(1), optimalParameters(2))
+  }
+
+  private def findAnomalies(
+      testSeries: Vector[Double],
+      forecasts: Seq[Double],
+      startIndex: Int,
+      residualSD: Double)
+    : Seq[(Int, Anomaly)] = {
+
+    testSeries.zip(forecasts).zipWithIndex.collect {
+      case ((inputValue, forecastedValue), detectionIndex)
+        if math.abs(inputValue - forecastedValue) > 1.96 * residualSD =>
```

Comment:
should this threshold be configurable?

Reply:
Yep should be, though I intentionally didn't expose that for now to keep the constructor as easy to understand as possible.

Let's make it configurable once we have an actual use-case for the sensitivity of the detection. What do you think?